assert/assertions.go

```diff
@@ -302,6 +304,10 @@ func IsType(t TestingT, expectedType interface{}, object interface{}, msgAndArgs
 // Pointer variable equality is determined based on the equality of the
 // referenced values (as opposed to the memory addresses).
 func Equal(t TestingT, expected, actual interface{}, msgAndArgs ...interface{}) bool {
+	if err := validateEqualArgs(expected, actual); err != nil {
+		return Fail(t, fmt.Sprintf("Invalid operation: %#v == %#v (%s)",
+			expected, actual, err), msgAndArgs...)
+	}
```

Comment:
people sometimes mix up expected and actual.

I would say it's better to fail if either `expected` or `actual` are a function.

People can compare functions using `.Nil` and `.NotNil`.